File: DSA/RAJAS-BAGGA/Day_5_Question_1.java
class Node {
    int data;
    Node next;
    public Node(int data) {this.data = data;}
}

class Day_5_Question_1 {
    public static void printReverseLinkedList(Node head) {
        if (head == null)
            return;
        
        printReverseLinkedList(head.next);
        System.out.printf("%d ", head.data);
    }
    
    public static void main(String[] args) {
        Node n1 = new Node(1);
        Node n2 = new Node(2);
        Node n3 = new Node(3);
        n1.next = n2;
        n2.next = n3;
        n3.next = null;
        printReverseLinkedList(n1);
    }
}

File: DSA/Ansh-Gulati/Day-5/Day-5_Question-1.cpp
#include <iostream>
using namespace std;

class Node
{
public:
    int data;
    Node *next;

    Node(int data)
    {
        this->data = data;
        this->next = nullptr;
    }
};

void printReverse(Node *head)
{
    if (head == nullptr)
    {
        return;
    }
    printReverse(head->next);
    cout << head->data << " ";
}

Node *createLinkedList()
{
    Node *head = nullptr;
    Node *tail = nullptr;
    int n;

    cout << "Enter the number of elements: ";
    cin >> n;

    cout << "Enter the elements: ";
    for (int i = 0; i < n; ++i)
    {
        int value;
        cin >> value;
        Node *newNode = new Node(value);
        if (head == nullptr)
        {
            head = newNode;
            tail = newNode;
        }
        else
        {
            tail->next = newNode;
            tail = newNode;
        }
    }
    return head;
}

void deleteLinkedList(Node *head)
{
    Node *temp;
    while (head != nullptr)
    {
        temp = head;
        head = head->next;
        delete temp;
    }
}

int main()
{
    Node *head = createLinkedList();
    cout << "The elements in reverse order are: ";
    printReverse(head);
    cout << endl;

    // Clean up memory
    deleteLinkedList(head);

    return 0;
}
